### PR TITLE
Update to Java 11 and separate uberjars

### PR DESCRIPTION
**Pull request by ThadHouse**
Matches shuffleboard.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -1,131 +1,137 @@
...
+
+// Ensure that the WPILibVersioningPlugin is setup by setting the release type, if releaseType wasn't
+// already specified on the command line
+if (!hasProperty("releaseType")) {
+    WPILibVersion {
+        releaseType = ReleaseType.DEV
+    }
+}
+
+// Only load the project version once, then share it
```
Comment (SamCarlberg) [inline build.gradle.kts]: Not relevant for a single-project build like OutlineViewer

File: build.gradle.kts [hunk 2]
```diff
@@ -1,131 +1,137 @@
...
+val projectVerion = getWPILibVersion()
+
+version = projectVerion
+
+val theMainClassName = "edu.wpi.first.outlineviewer.Main"
+
+// Note: plugins should override this
```
Comment (SamCarlberg) [inline build.gradle.kts]: No plugin projects

File: build.gradle.kts [hunk 3]
```diff
@@ -1,131 +1,137 @@
...
 }
 
-apply {
-    plugin("pmd")
-    plugin("findbugs")
-    plugin("jacoco")
-    plugin("org.junit.platform.gradle.plugin")
+application {
+    mainClassName = theMainClassName
+    applicationDefaultJvmArgs = listOf(
```
Comment (SamCarlberg) [inline build.gradle.kts]: Not really useful for OutlineViewer

File: build.gradle.kts [hunk 4]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
-    jacoco {
-        applyTo(junitPlatformTest)
-    }
-    task<JacocoReport>("jacocoJunit5TestReport") {
-        executionData(junitPlatformTest)
-        sourceSets(java.sourceSets["main"])
-        sourceDirectories = files(java.sourceSets["main"].allSource.srcDirs)
-        classDirectories = files(java.sourceSets["main"].output)
+
+tasks.withType<Test> {
```
Comment (SamCarlberg) [inline build.gradle.kts]: Maybe `tasks.withType<Test>().configureEach {` ?

/cc @JLLeitschuh

File: build.gradle.kts [hunk 5]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
-                 * but not with the virtualized ones in headless mode.
-                 */
-                exclude("NonHeadlessTests")
-            }
-        }
+tasks.withType<Javadoc> {
+    isFailOnError = false
+}
+
+dependencies {
```
Comment (SamCarlberg) [inline build.gradle.kts]: Empty dependencies block?

File: build.gradle.kts [hunk 6]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
     publications {
-        create<MavenPublication>("OutlineViewer") {
-            artifactId = "OutlineViewer"
-            getWPILibVersion()?.let { version = it }
-            val shadowJar: ShadowJar by tasks
-            artifact (shadowJar) {
-                classifier = null
+        create<MavenPublication>("app") {
+            groupId = "edu.wpi.first.outlineviewer"
+            artifactId = "outlineviewer"
```
Comment (SamCarlberg) [inline build.gradle.kts]: Do we want to change the name of the JAR files?

File: build.gradle.kts [hunk 7]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
+
 publishing {
     publications {
-        create<MavenPublication>("OutlineViewer") {
-            artifactId = "OutlineViewer"
-            getWPILibVersion()?.let { version = it }
-            val shadowJar: ShadowJar by tasks
-            artifact (shadowJar) {
-                classifier = null
+        create<MavenPublication>("app") {
```
Comment (SamCarlberg) [inline build.gradle.kts:208]: Not sure what difference this makes, if any

File: build.gradle.kts [hunk 8]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
+            groupId = "edu.wpi.first.outlineviewer"
+            artifactId = "outlineviewer"
+            version = project.version as String
+            nativeShadowTasks.forEach {
+                artifact(it) {
+                    classifier = it.classifier
+                }
             }
+            artifact(sourceJar)
+            artifact(javadocJar)
```
Comment (SamCarlberg) [inline build.gradle.kts]: I don't think we need to bother publishing sources and javadocs for a standalone application like OutlineViewer

File: build.gradle.kts [hunk 9]
```diff
@@ -134,90 +140,113 @@ tasks.withType<JacocoReport> {
...
     }
 }
 
-// Ensure that the WPILibVersioningPlugin is setup by setting the release type, if releaseType wasn't
-// already specified on the command line
-if (!hasProperty("releaseType")) {
-    WPILibVersion {
-        releaseType = ReleaseType.DEV
-    }
+dependencies {
```
Comment (SamCarlberg) [inline build.gradle.kts]: Empty dependencies block?

File: src/main/java/edu/wpi/first/outlineviewer/Main.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+ * which is not in a package exported by the {@code javafx.graphics} module.
+ */
+@SuppressWarnings("PMD.UseUtilityClass") // Nope.
+public final class Main {
+  @SuppressWarnings("JavadocMethod")
+  public static void main(String[] args) {
+    // JavaFX 11+ uses GTK3 by default, and has problems on some display servers
+    // This flag forces JavaFX to use GTK2
+    System.setProperty("jdk.gtk.version", "2");
+    LauncherImpl.launchApplication(OutlineViewer.class, args);
```
Comment (SamCarlberg) [inline src/main/java/edu/wpi/first/outlineviewer/Main.java]: I'd call `ApplicationLaunch(OutlineViewer.class, args);` since we don't have a preloader. Shuffleboard has a preloader, which is why it uses `LauncherImpl` directly

**Review by SamCarlberg — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 7 above)
Comment (ThadHouse) [reply build.gradle.kts:208]: The name there does not matter to what actually gets published

File: build.gradle.kts (hunk 6 above)
Comment (ThadHouse) [reply build.gradle.kts]: I reverted this

File: build.gradle.kts [hunk 11]
```diff
@@ -1,223 +1,236 @@
...
-    application
     `maven-publish`
     jacoco
     java
     checkstyle
+    application
     pmd
-    id("edu.wpi.first.wpilib.versioning.WPILibVersioningPlugin") version "2.0"
+    id("edu.wpi.first.wpilib.versioning.WPILibVersioningPlugin") version "2.2"
     id("com.github.johnrengelman.shadow") version "2.0.1"
```
Comment (AustinShalit) [inline build.gradle.kts]: Update to 2.0.4

File: build.gradle.kts [hunk 12]
```diff
@@ -1,223 +1,236 @@
...
-    configFile = file("$rootDir/checkstyle.xml")
-    toolVersion = "8.1"
+    toolVersion = "8.12"
 }
 
 pmd {
+    toolVersion = "6.7.0"
     isConsoleOutput = true
-    sourceSets = setOf(java.sourceSets["main"], java.sourceSets["test"])
+    sourceSets = setOf(java.sourceSets["main"])
```
Comment (AustinShalit) [inline build.gradle.kts]: Why this change.  We should run on main and test.

File: build.gradle.kts [hunk 13]
```diff
@@ -1,223 +1,236 @@
-
 import edu.wpi.first.wpilib.versioning.ReleaseType
 import org.gradle.api.Project
-import org.gradle.api.plugins.quality.FindBugs
```
Comment (AustinShalit) [inline build.gradle.kts:4]: Do not remove.  Replace with spot bugs.  See #74

File: build.gradle.kts [hunk 14]
```diff
@@ -228,35 +241,12 @@ fun Project.`checkstyle`(configure: org.gradle.api.plugins.quality.CheckstyleExt
 /**
  * Retrieves the [pmd][org.gradle.api.plugins.quality.PmdExtension] project extension.
  */
-val Project.`pmd`: org.gradle.api.plugins.quality.PmdExtension get() =
-    extensions.getByName("pmd") as org.gradle.api.plugins.quality.PmdExtension
+val Project.`pmd`: org.gradle.api.plugins.quality.PmdExtension
+    get() =
+        extensions.getByName("pmd") as org.gradle.api.plugins.quality.PmdExtension
```
Comment (AustinShalit) [inline build.gradle.kts]: I am not sure you need any of these anymore.  See #74.

**Review by AustinShalit — CHANGES_REQUESTED** [review]
Why all of the changes to the `checkstyle.xml`?

**Comment by ThadHouse** [comment]
The checkstyle.xml changes were because I copied it from the shuffleboard build, because the one currently here was failing with the config error, and that was the only way I found to fix.

File: build.gradle.kts (hunk 11 above)
Comment (ThadHouse) [reply build.gradle.kts]: Fixed

File: build.gradle.kts (hunk 13 above)
Comment (ThadHouse) [reply build.gradle.kts:4]: Fixed

**Comment by ThadHouse** [comment]
Fixed all the review things, so this should be good to go.

**Review by AustinShalit — CHANGES_REQUESTED** [review]
`./gradlew check` runs no tests for me

**Comment by ThadHouse** [comment]
It does. What you're probably seeing is all the UI tests are disabled because of testfx issues. I made it match how shuffleboard was done, where I marked everything UI with `@Tag("UI")`. But if you check the build log it does run some.

File: buildSrc/src/main/kotlin/JavaFXExtensions.kt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Generates dependencies for all platform-specific configurations on a JavaFX artifact. The artifact name is
+ * prepended with `"javafx-"`, so `name` should be something like `"base"` instead of `"javafx-base"` or `"graphics"`
+ * vs `"javafx-graphics"`.
+ */
+fun DependencyHandler.javafx(name: String) {
+    forEachPlatform {
+        javafx(name = name, platform = it)
+    }
+}
```
Comment (JLLeitschuh) [inline buildSrc/src/main/kotlin/JavaFXExtensions.kt:26]: @SamCarlberg Do you want to turn this stuff into a gradle plugin??

Comment (ThadHouse) [reply buildSrc/src/main/kotlin/JavaFXExtensions.kt:26]: In slack we said yes, but we won't have time before merging this, as we need this merged today for a beta release on friday.

Comment (JLLeitschuh) [reply buildSrc/src/main/kotlin/JavaFXExtensions.kt:26]: Okay.

**Review by JLLeitschuh — DISMISSED** [review]
LGTM!

**Comment by ThadHouse** [comment]
This closes #72 as well.

**Comment by ThadHouse** [comment]
Supersedes #74
